dag_in_context/src/schema_helpers.rs

```diff
@@ -206,6 +233,166 @@ impl Expr {
+        let arg_ctx = arg.get_ctx();
+        Self::subst_with_cache(arg, &arg_ty, arg_ctx, within, &mut subst_cache)
+    }
+
+    // If arg_ctx exists, will also substitute context. Otherwise, old context is maintained.
+    fn subst_with_cache(
+        arg: &RcExpr,
+        arg_ty: &Type,
+        arg_ctx: Option<&Assumption>,
+        within: &RcExpr,
+        subst_cache: &mut HashMap<*const Expr, RcExpr>,
+    ) -> RcExpr {
+        if let Some(substed) = subst_cache.get(&Rc::as_ptr(within)) {
+            return substed.clone();
+        }
+
+        let substed = match within.as_ref() {
+            // Substitute!
+            Expr::Arg(_) => arg.clone(),
+
+            // Propagate through current scope
+            Expr::Top(op, x, y, z) => Rc::new(Expr::Top(
+                op.clone(),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, y, subst_cache),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, z, subst_cache),
+            )),
+            Expr::Bop(op, x, y) => Rc::new(Expr::Bop(
+                op.clone(),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, y, subst_cache),
+            )),
+            Expr::Uop(op, x) => Rc::new(Expr::Uop(
+                op.clone(),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+            )),
+            Expr::Get(x, i) => Rc::new(Expr::Get(
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+                *i,
+            )),
+            Expr::Alloc(amt, x, y, ty) => Rc::new(Expr::Alloc(
+                *amt,
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, y, subst_cache),
+                ty.clone(),
+            )),
+            Expr::Call(name, x) => Rc::new(Expr::Call(
+                name.clone(),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+            )),
+            Expr::Single(x) => Rc::new(Expr::Single(Self::subst_with_cache(
+                arg,
+                arg_ty,
+                arg_ctx,
+                x,
+                subst_cache,
+            ))),
+            Expr::Concat(x, y) => Rc::new(Expr::Concat(
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, x, subst_cache),
+                Self::subst_with_cache(arg, arg_ty, arg_ctx, y, subst_cache),
+            )),
+            Expr::If(pred, input, then, els) => {
+                let new_pred = Self::subst_with_cache(arg, arg_ty, arg_ctx, pred, subst_cache);
+                let new_input = Self::subst_with_cache(arg, arg_ty, arg_ctx, input, subst_cache);
+                Rc::new(Expr::If(
+                    new_pred.clone(),
+                    new_input.clone(),
+                    then.clone(),
```

Comment:
Using `replace_ctx` for if/switch branches or loop bodies is much slower. To generate egglog for `fib_recursive.bril`, it's < 1 s without replacing vs. > 1 min with replacing. This is because replacing context creates more calls to inline. Previously-inlined calls in branches and loop bodies are not equivalent to the context-replaced calls, so the context-replaced calls will also have their corresponding bodies substituted and added to the inlining unions.

If it is still correct to not replace the context, I think it is better to not replace the context and make inlining run in a reasonable time at the expense of losing some opportunities for optimization.